app/models/log_event.rb

```diff
@@ -3,8 +3,92 @@ class LogEvent < ActiveRecord::Base
   belongs_to :user
   belongs_to :loggable, polymorphic: true
 
-  def self.log(loggable, event_type, user)
-    create(loggable: loggable, event_type: event_type, user_id: user.id)
+  def self.log(loggable, event_type, event_time, user)
+    create(
+      loggable: loggable, event_type: event_type,
+      event_time: event_time, user_id: user.try(:id))
+  end
+
+  def self.search(start_date: nil, end_date: nil, events: [], query: nil)
+    LogEventSearcher.new(
+      start_date: start_date, end_date: end_date, events: events, query: query).search
+  end
+
+  def locale_tag
+    "#{loggable_type.underscore.downcase}.#{event_type}"
+  end
+
+  def loggable_to_s
+    case loggable
+    when AccountUser
+      "#{loggable.account} / #{loggable.user}"
+    else
+      loggable.to_s
+    end
+  end
+
+  class LogEventSearcher
+
+    ALLOWED_EVENTS = ["account.create", "account.update",
+                      "account_user.create", "account_user.delete",
+                      "user.create"].freeze
+
+    attr_accessor :start_date, :end_date, :events, :query
+
+    def initialize(start_date: nil, end_date: nil, events: [], query: nil)
+      @start_date = start_date
+      @end_date = end_date
+      @events = filter_events(events)
+      @query = query
+    end
+
+    def filter_events(events)
+      return events if events.blank?
+      events.select { |event| event.in?(ALLOWED_EVENTS) }
+    end
+
+    def search
+      result = LogEvent.all
+      result = result.merge(filter_date) if start_date || end_date
+      result = result.merge(filter_event) if events.present?
+      result = result.merge(filter_query) if query.present?
+      result
+    end
+
+    def dates
+      [(start_date || 10.years.ago), (end_date || Date.current)]
```

Comment:
It is arbitrary, the obvious other choice would be `LogEvent.minimum(:event_time)`.